packages/typespec-client-generator-core/src/package.ts

```diff
@@ -0,0 +1,734 @@
+  getClientNamespaceStringHelper,
+  getDocHelper,
+  getHashForType,
+  getSdkTypeBaseHelper,
+  isAcceptHeader,
+  updateWithApiVersionInformation,
+} from "./internal-utils.js";
+import {
+  getClientNamespaceString,
+  getDefaultApiVersion,
+  getLibraryName,
+  getPropertyNames,
+} from "./public-utils.js";
+import {
+  addEncodeInfo,
+  addFormatInfo,
+  getAllModels,
+  getClientType,
+  getSdkCredentialParameter,
+  getSdkModelPropertyType,
+} from "./types.js";
+
+function getSdkHttpBodyParameters(
+  context: SdkContext<SdkHttpOperation>,
+  httpOperation: HttpOperation,
+  methodParameters: SdkParameter[]
+): SdkBodyParameter[] | undefined {
+  const tspBody = httpOperation.parameters.body;
+  if (tspBody === undefined) return undefined;
+  let contentTypes = tspBody.contentTypes;
+  if (contentTypes.length === 0) {
+    contentTypes = ["application/json"];
+  }
+  const defaultContentType = contentTypes.includes("application/json")
+    ? "application/json"
+    : contentTypes[0];
+  if (!tspBody.parameter) {
+    const bodyType = getClientType(context, tspBody.type, httpOperation.operation);
+    return [
+      {
+        kind: "body",
+        nameInClient: "body",
+        description: getDocHelper(context, tspBody.type).description,
+        details: getDocHelper(context, tspBody.type).details,
+        onClient: false,
+        contentTypes: contentTypes,
+        defaultContentType: defaultContentType,
+        isApiVersionParam: false,
+        apiVersions: getAvailableApiVersions<SdkHttpOperation>(context, tspBody.type),
+        type: bodyType,
+        optional: false,
+      },
+    ];
+  }
+  const body = getSdkModelPropertyType(context, tspBody.parameter!, {
+    operation: httpOperation.operation,
+    defaultContentType: defaultContentType,
+  });
+  const methodBodyParameter = methodParameters.find(
+    (x) => x.nameInClient === getPropertyNames(context, tspBody.parameter!)[0]
+  );
+  if (body.kind !== "body") throw new Error("blah");
+  if (methodBodyParameter) {
+    return [
+      {
+        ...body,
+        contentTypes,
+        defaultContentType,
+      },
+    ];
+  } else {
+    // this means that the body parameter is a property on one of the method parameters
+    for (const methodParameter of methodParameters) {
+      if (methodParameter.type.kind === "model") {
+        const bodyProperty = methodParameter.type.properties.find((x) => x.kind === "body");
+        if (bodyProperty) {
+          return [
+            {
+              ...body,
+              contentTypes,
+              defaultContentType,
+            },
+          ];
+        }
+      }
+    }
+  }
+  throw new Error("blah");
+}
+
+function createContentTypeOrAcceptHeader(
+  context: SdkContext<SdkHttpOperation>,
+  bodyObject: SdkBodyParameter | SdkHttpResponse
+): Omit<SdkMethodParameter, "kind"> {
+  const nameInClient = bodyObject.kind === "body" ? "contentType" : "accept";
+  let type: SdkType = {
+    kind: "string",
```

Comment:
Shouldn't this be `kind: "constant"`? I'm seeing the `accept` header show up as parameter due to its non-constant nature.

Reply:
Ok maybe not here as we need to handle multiple content types. But it should be `constant` for the `accept` header.

Reply:
To clarify my first point, I see `accept` show up as non-constant for things like `application/octet-stream` and `image/png`. Removing the `json` constraint below resolves the issue. But is that correct?

Reply:
That's a good point, just because we don't default `contentType` bc of not having a clear default, doesn't mean we shouldn't default `accept`, bc adding an `accept` new mimetype would be breaking

Reply:
Fixed in b8756c8.